### PR TITLE
Simplify recycler design

### PR DESCRIPTION
**Pull request by ninja-**
Doesn't seem to affect benchmarks but looks a lot better(without arrays). wdyt about this change?
I am posting this just as an idea to test.

### CODE REVIEW

**Comment by ninja-** [comment]
all the failed tests are *endian* tests but I doubt it's related...

**Comment by normanmaurer** [comment]
@ninja- I think we should first proof that your code is better or at least as good as it was before. The current recycler code was tuned a lot to ensure it performs as good as possible when used concurrently. So

**Comment by ninja-** [comment]
well I mentioned it doesn't seem to change benchmarks so

**Comment by normanmaurer** [comment]
@ninja- how did you verify this ?

**Comment by ninja-** [comment]
@normanmaurer changed RecyclableArrayListBenchmark to do more stuff and then run the benchmarks on both versions

**Comment by normanmaurer** [comment]
hmm... I need to think about it more. Also do the tests all pass locally ?

File: common/src/main/java/io/netty/util/Recycler.java [hunk 1]
```diff
@@ -85,7 +84,7 @@ protected Recycler() {
     }
 
     protected Recycler(int maxCapacity) {
-        this.maxCapacity = Math.max(0, maxCapacity);
+        this.maxCapacity = Math.min(0, maxCapacity);
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/Recycler.java]: why this change ? This makes no sense as this will always result in 0 or a negative value

File: common/src/test/java/io/netty/util/RecyclerTest.java [hunk 2]
```diff
@@ -24,7 +24,7 @@
 
 public class RecyclerTest {
 
-    @Test(expected = IllegalStateException.class)
+    //@Test(expected = IllegalStateException.class)
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/RecyclerTest.java]: Why this change ?

File: common/src/main/java/io/netty/util/Recycler.java (hunk 1 above)
Comment (ninja-) [inline common/src/main/java/io/netty/util/Recycler.java]: sorry this was a dumb mistake, I aimed to fix it on the other line while on the web editor

File: common/src/test/java/io/netty/util/RecyclerTest.java (hunk 2 above)
Comment (ninja-) [inline common/src/test/java/io/netty/util/RecyclerTest.java]: I removed the "already recycled" checks but could leave them untouched as well

Comment (normanmaurer) [inline common/src/test/java/io/netty/util/RecyclerTest.java]: yes these should not be removed as this is clearly an error.

Comment (ninja-) [inline common/src/test/java/io/netty/util/RecyclerTest.java]: why? these checks don't matter that much if user code is correct, the same way you removed ensureAccessible checks

Comment (normanmaurer) [inline common/src/test/java/io/netty/util/RecyclerTest.java]: Because user could may not be correct?

Comment (ninja-) [inline common/src/test/java/io/netty/util/RecyclerTest.java]: but the same way user may not be correct and operate on dead buffers without ensureAccessible...so maybe a setting as usual

Comment (normanmaurer) [inline common/src/test/java/io/netty/util/RecyclerTest.java]: @ninja- first I would like a proof that this actually is a performance bottleneck.

Comment (ninja-) [inline common/src/test/java/io/netty/util/RecyclerTest.java]: @normanmaurer but is there _any_ bottleneck at all?

Comment (normanmaurer) [inline common/src/test/java/io/netty/util/RecyclerTest.java]: @ninja- maybe I was not clear... If you consider removing safety checks I would like to get a proof that it is actually such a big deal that it worth it. If there is no notable performance gain (which I suspect here) I don't think it makes sense to allow to remove the safety checks.

Comment (ninja-) [inline common/src/test/java/io/netty/util/RecyclerTest.java]: @normanmaurer a boolean recycled=true would do or there is a reason to use these ids?

Comment (johnou) [inline common/src/test/java/io/netty/util/RecyclerTest.java]: @ninja- pretty sure @normanmaurer would rather you leave all the safety checks in place (even without a property to disable them) unless you can provide benchmark results showing that with the safety checks disabled you see a notable performance gain..

Perhaps benchmark results between old recycler and the "simplified recycler" and then finally benchmark results between the "simplified recycler" and the "simplified recycler" with safety checks disabled.
